### PR TITLE
Parse percentage encoded paths

### LINKED ISSUE #134: Can not parse uri
can not parse uris `/echo/abcdefgh_i-j%20/abcdefg_i-j%20478` or `/echo/abcdefgh_i-j /abcdefg_i-j 478`

### PR DESCRIPTION
**Pull request by carllerche**
Fixes #134

### CODE REVIEW

File: src/uri.rs [hunk 1]
```diff
@@ -1200,12 +1230,30 @@ impl PathAndQuery {
...
-        for i in 0..src.len() {
+        let mut i = 0;
+
+        while i < src.len() {
             let b = src[i];
 
             match URI_CHARS[b as usize] {
                 0 => {
-                    return Err(ErrorKind::InvalidUriChar.into());
+                    if b == b'%' {
```
Comment (seanmonstar) [inline src/uri.rs]: Could this just follow the pattern of `?` and `#` instead of matching 0 and then checking again what byte it is?

Comment (carllerche) [reply src/uri.rs]: `URI_CHARS` was also used in other situations where `%` isn't permitted. It seemed roughly equivalent to do it this way vs. create `PATH_CHARS` that permitted `%`.

Comment (seanmonstar) [reply src/uri.rs]: Oh, OK!

**Comment by nox** [comment]
Which spec is this patch following? It seems that now Hyper cannot cope with a URL containing an unencoded `|` char in its query.

**Pull request merged**